rfcs/0074-tvmscript-unified-printer.md

```diff
@@ -0,0 +1,487 @@
+//    It's optional and represents the name of IR
+// 3. The first argument to the print function is the node to be printed
+// 4. The second argument is instance of `IRDocsifier`, which can be used
+//    to recursively translate the child nodes.
+// 5. The print method returns a subclass of Doc
+
+TVM_STATIC_IR_FUNCTOR(IRDocsifier, vtable).set_dispatch<Range>([](Range e, IRDocsifier p) {
+  return SliceDoc(p->AsExprDoc(e->min), p->AsExprDoc(e->min + e->extent));
+});
+
+// The first arg to the `set_dispatch` can be omitted, and 
+// the print function will be registered the default layer.
+// It will be called by default and can be overriden by registering
+// another print function under an IR name. For example,
+
+TVM_STATIC_IR_FUNCTOR(IRDocsifier, vtable).set_dispatch<Range>("relax", [](Range e, IRDocsifier p) {
+  ...
+});
+
+// The function above will be called instead of the previous one, 
+// if Printer is printing relax.
+```
+
+- Dispatch
+
+```cpp
+auto tir_dispatch_ctx = ir_docsifier->WithDispatchToken("tir");
+Doc doc = ir_docsifier->AsDoc<Doc>(node);
+
+// Here we setup the ir_docsifier to call print functions under 
+// the 'tir' dispatch token, and then call the AsDoc method to 
+// translate `node`, as an ObjectRef, into `Doc`, by using the 
+// print functions registered in the dispatch table.
+
+template <class TDoc>
+TDoc AsDoc(const ObjectRef& obj) const {
+  return Downcast<TDoc>(AsDocImpl(obj));
+}
+```
+
+## Frame Spec
+
+Frame provides the contextual information during printing. Most commonly, frame
+contains variable defined in the current scope (like tir function, tir block,
+tir loop). A subclass of Frame can be created to store more specific
+information. For instance, `TIRLoopFrame` should contain the information about
+the TIR for loop in order to print iter var remapping when printing
+BlockRealize.
+
+```cpp
+class FrameNode : public Object {
+ public:
+  Array<ObjectRef> objs;
+  SymbolTableNode* sym;
+
+  /*!
+   * \brief Set the name of a variable IR node
+   */
+	virtual IdDoc DefByName(const ObjectRef& obj, const String& name);
+  /*!
+   * \brief Set the doc of a variable IR node
+   * \details This is useful when the variable is implicitly defined in the TVMScript.
+   *          For example, when defining a `tir::Buffer buf`, buf->data is also a tir::Var,
+   *          which should be printed as `buf.data`, rather than an identifier
+   *          in the TVMScript.
+   */
+  virtual ExprDoc DefByDoc(const ObjectRef& obj, const ExprDoc& doc);
+}
+```
+
+## Folder Structure
+
+```
+src/script/printer/
+├── doc.cc
+├── doc.h
+├── doc_printer.cc
+├── doc_printer.h
+├── frame.cc
+├── frame.h
+├── functor.h
+├── ir_docsifier.cc
+├── ir_docsifier.h
+├── symbol_table.cc
+├── symbol_table.h
+├── tir 
+│   ├── buffer.cc
+│   ├── expr.cc
+│   ├── function.cc
+│   ├── op.cc
+│   ├── stmt.cc
+│   ├── tir.h
+│   ├── type.cc
+│   ├── utils.h
+│   └── var.cc
+├── tvmscript_unified_printer.cc
+├── util.cc
+└── util.h
```

Comment:
A couple of ideas:
- public header files belong to `include/tvm/script/printer` instead of the private `src` folder
- no need to include every detail, but only deliver the core idea.
- it's always better to annotate directly in the figure instead of explaining later in text.

For example:

```
src/script/printer/
├── core  # core infra, which is IR-agnostic
├── tir # TIR dialect
├── relax # hypothetical Relax dialect (not part of our RFC)
```

Reply:
Updated